pkgpanda/actions.py

```diff
@@ -0,0 +1,293 @@
+import collections
+import os
+import sys
+from functools import partial
+from subprocess import check_call
+
+from pkgpanda import PackageId, requests_fetcher
+from pkgpanda.exceptions import FetchError, PackageError, ValidationError
+from pkgpanda.util import (extract_tarball, if_exists, load_json, load_string,
+                           write_string)
+
+
+DCOS_TARGET_CONTENTS = """[Install]
+WantedBy=multi-user.target
+"""
+
+
+def activate_packages(install, repository, package_ids, systemd=True, block_systemd=False):
+    """Replace the active package set with package_ids.
+
+    install: pkgpanda.Install
+    repository: pkgpanda.Repository
+    package_ids: sequence of package IDs to activate
+    systemd: start/stop systemd services (default: True)
+    block_systemd: if systemd, block waiting for systemd services to come up (default: False)
+
+    """
+    assert isinstance(package_ids, collections.Sequence)
+    try:
+        install.activate(repository.load_packages(package_ids))
+        if systemd:
+            _start_dcos_target(block_systemd)
+    except ValidationError as ex:
+        print("Validation Error: {0}".format(ex))
+        sys.exit(1)
+    except PackageError as ex:
+        print("Package Error: {0}".format(ex))
+
+
+def swap_active_package(install, repository, package_id, systemd=True, block_systemd=False):
+    """Replace an active package with a package_id with the same name.
+
+    swap(install, repository, 'foo--version') will replace the active 'foo'
+    package with 'foo--version'.
+
+    install: pkgpanda.Install
+    repository: pkgpanda.Repository
+    package_id: package ID to activate
+    systemd: start/stop systemd services (default: True)
+    block_systemd: if systemd, block waiting for systemd services to come up (default: False)
+
+    """
+    active = install.get_active()
+    # TODO(cmaloney): I guarantee there is a better way to write this and
+    # I've written the same logic before...
+    packages_by_name = dict()
+    for id_str in active:
+        pkg_id = PackageId(id_str)
+        packages_by_name[pkg_id.name] = pkg_id
+
+    new_id = PackageId(package_id)
+    if new_id.name not in packages_by_name:
+        print("ERROR: No package with name {} currently active to swap with.".format(new_id.name))
+
+    packages_by_name[new_id.name] = new_id
+    new_active = list(map(str, packages_by_name.values()))
+    # Activate with the new package name
+    activate_packages(install, repository, new_active, systemd, block_systemd)
+
+
+def fetch_packages(repository, repository_url, package_ids, work_dir=None):
+    """Fetch package_ids from repository_url into repository.
+
+    repository: pkgpanda.Repository
+    repository_url: URL for remote package repository
+    package_ids: sequence of package IDs to fetch
+    work_dir: location for temporary files (default: current working dir)
+
+    """
+    assert isinstance(package_ids, collections.Sequence)
+    work_dir = work_dir or os.getcwd()
```

Comment:
If this is going to be a library, we can't / don't want to assume os.getcwd(). work_dir needs to be passed in as a parameter.